### PR TITLE
Feature/gradients

### PR DESCRIPTION
**Pull request by meganrm**
Problem
=======
#120 

Solution
========
Adjusted the old gradient system to work with our new setup 
Created new function that calculates the distance from a surface to the next bounding surface. 

## Type of change
Please delete options that are not relevant.

* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. `pack -r cellpack/test-recipes/v2/peroxisomes_gradient.json -c packing-configs/run.json`

Screenshots (optional):
-----------------------
Screen Shot 2022-11-22 at 9 56 46 AM

### CODE REVIEW

File: cellpack/autopack/Environment.py [hunk 1]
```diff
@@ -280,6 +280,23 @@ def _setup(self):
                 self.referenced_objects,
             ) = Recipe.resolve_composition(self.recipe_data)
             self.create_objects()
+        if self.use_gradient:
+            gradients = self.recipe_data["gradients"]
```
Comment (prismofeverything) [inline cellpack/autopack/Environment.py:284]: Only mildly redundant to have a boolean `use_gradient` here along with the presence or absence of a gradient in `self.recipe_data["gradients"]`. Could we collapse this check into a check for existence in recipe_data?

File: cellpack/autopack/Environment.py [hunk 2]
```diff
@@ -280,6 +280,23 @@ def _setup(self):
...
+                    )
+                self.set_gradient(gradient_name, gradient_data)
+
+    def get_compartment_object_by_name(self, compartment_name):
+        """
+        Returns compartment object by name
+        """
+        for compartment in self.compartments:
+            if compartment.name == compartment_name:
+                return compartment
```
Comment (prismofeverything) [inline cellpack/autopack/Environment.py:299]: This is essentially reimplementing a dictionary lookup as a linear search.... I know we never have that many compartments but maybe we will some day? How plausible is it to just make `self.compartments` a dictionary? If you need the list you could just `self.compartments.values` etc.

File: cellpack/autopack/Environment.py [hunk 3]
```diff
@@ -684,19 +698,16 @@ def sortIngredient(self, reset=False):
             if o.surfaceRecipe:
                 o.surfaceRecipe.sort()
 
-    def setGradient(self, **kw):
+    def set_gradient(self, name, gradient_data):
```
Comment (prismofeverything) [inline cellpack/autopack/Environment.py:701]: Thanks for removing `**kw` here : )

File: cellpack/autopack/Gradient.py [hunk 4]
```diff
@@ -73,7 +73,17 @@ def __init__(self, name, mode="X", description="", direction=None, bb=None, **kw
...
+            "X",
+            "Y",
+            "Z",
+            "-X",
+            "-Y",
+            "-Z",
+            "direction",
+            "radial",
+            "surface",
+        ]
```
Comment (prismofeverything) [inline cellpack/autopack/Gradient.py:86]: I wonder about these modes.... they could in general have associated parameters or something ("radius" for radial?). Could be worthwhile to turn these modes into objects with attributes at some point.

File: cellpack/autopack/Gradient.py [hunk 5]
```diff
@@ -217,189 +228,121 @@ def pickPoint(self, listPts):
...
-        xl, yl, zl = bb[0]
-        xr, yr, zr = bb[1]
+        center = self.direction
+        max_distance = self.radius
+        distances = get_distances_from_point(master_grid_positions, center)
+        self.distances = (
+            numpy.where(distances < max_distance, distances, max_distance)
+            / max_distance
+        )
+        self.set_weights_by_mode()
```
Comment (prismofeverything) [inline cellpack/autopack/Gradient.py:288]: Great to see more numpy optimizations work their way in.

**Review by prismofeverything — APPROVED** [review]
Hey @meganrm looks great! Just a few comments but overall I think this is ready to go.

File: cellpack/autopack/Environment.py [hunk 6]
```diff
@@ -2764,7 +2786,7 @@ def setupPanda(self):
             self.rb_panda = []
         for o in self.compartments:
             if o.rbnode is None:
-                o.rbnode = o.addShapeRB()  # addMeshRBOrganelle(o)
+                o.rbnode = o.addShapeRB(self)  # addMeshRBOrganelle(o)
```
Comment (mogres) [inline cellpack/autopack/Environment.py:2789]: Could we update `o` to something more descriptive?

File: cellpack/autopack/Environment.py (hunk 2 above)
Comment (meganrm) [reply cellpack/autopack/Environment.py:299]: at the moment we're mostly always iterating over all compartments so I think it makes sense to have it be an array

**Pull request merged**